channel/channel.go

```diff
@@ -0,0 +1,166 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package channel
+
+import (
+	"github.com/megaease/easeprobe/notify"
+	"github.com/megaease/easeprobe/probe"
+	log "github.com/sirupsen/logrus"
+	"sync/atomic"
+)
+
+const kind = "channel"
+
+// Channel implements a config for Channel
+type Channel struct {
+	Name      string                    `yaml:"name"`      // unique name
+	Probers   map[string]*probe.Prober  `yaml:"probers"`   // probers
+	Notifiers map[string]*notify.Notify `yaml:"notifiers"` // notifiers
+	isWatch   int32                     `yaml:"-"`         // is watch
+	done      chan bool                 `yaml:"-"`         // done channel
+	channel   chan probe.Result         `yaml:"-"`         // notify channel
+
+}
+
+// NewEmpty creates a new empty Channel object with nil channel
+// After setup the probers, You have to call Config() to create the channel
+func NewEmpty(name string) *Channel {
+	return &Channel{
+		Name:      name,
+		Probers:   map[string]*probe.Prober{},
+		Notifiers: map[string]*notify.Notify{},
+		isWatch:   0,
+		done:      nil,
+		channel:   nil,
+	}
+}
+
+// Config configures the channel
+func (c *Channel) Config() {
+	c.done = make(chan bool)
+	c.channel = make(chan probe.Result, len(c.Probers))
+}
+
+// Done returns the done channel
+func (c *Channel) Done() chan bool {
+	return c.done
+}
+
+// Channel returns the notification channel
+func (c *Channel) Channel() chan probe.Result {
+	return c.channel
+}
+
+// Send sends the result to the channel
+func (c *Channel) Send(result probe.Result) {
+	c.channel <- result
+}
+
+// GetProber returns the Notify object
+func (c *Channel) GetProber(name string) *probe.Prober {
+	return c.Probers[name]
+}
+
+// SetProbers sets the Notify objects
+func (c *Channel) SetProbers(probers []*probe.Prober) {
+	for _, p := range probers {
+		c.SetProber(p)
+	}
+}
+
+// SetProber sets the Notify object
+func (c *Channel) SetProber(p *probe.Prober) {
+	if p == nil {
+		return
+	}
+	c.Probers[(*p).Name()] = p
+}
+
+// GetNotify returns the Notify object
+func (c *Channel) GetNotify(name string) *notify.Notify {
+	return c.Notifiers[name]
+}
+
+// SetNotifiers sets the Notify objects
+func (c *Channel) SetNotifiers(notifiers []*notify.Notify) {
+	for _, n := range notifiers {
+		c.SetNotify(n)
+	}
+}
+
+// SetNotify sets the Notify object
+func (c *Channel) SetNotify(n *notify.Notify) {
+	if n == nil {
+		return
+	}
+	c.Notifiers[(*n).GetName()] = n
+}
+
+var dryNotify bool
+
+// SetDryNotify sets the global dry run flag
+func SetDryNotify(dry bool) {
+	dryNotify = dry
+}
+
+// WatchEvent watches the notification event
+// Go through all of notification to notify the result.
+func (c *Channel) WatchEvent() {
+
+	// check if the channel is watching
+	if atomic.LoadInt32(&(c.isWatch)) != 0 {
```

Comment:
Seems this line should use `atomic.CompareAndSwapInt32`.

Reply:
I think you find a concurrent issue here. let me think about how to fix it.